blog/_posts/2025-02-28-afd-aks-ingress-tls-approuting.md

```diff
@@ -0,0 +1,414 @@
+az network vnet create \
+-g $RG \
+-n $VNET_NAME \
+--address-prefix $VNET_ADDRESS_SPACE \
+--subnet-name aks \
+--subnet-prefix $AKS_SUBNET_ADDRESS_SPACE 
+
+# Get a subnet resource ID, which we'll need when we create the AKS cluster
+VNET_SUBNET_ID=$(az network vnet subnet show -g $RG --vnet-name $VNET_NAME -n aks -o tsv --query id)
+```
+
+## Cluster Creation
+
+Now, let's create the AKS cluster. This will be a very plain AKS cluster, for simplicity, however we will deploy to our above created subnet and will enable the following features:
+
+- *AKS App Routing:* This will deploy managed Nginx Ingress
+- *App Routing Default Nginx Controller:* This allows us to ensure that the default deployment of App Routing uses a private IP
+- *Key Vault CSI Driver:* The Key Vault CSI driver will be used by App Routing to access our certificate in Azure Key Vault.
+
+```bash
+# NOTE: Make sure you give your cluster a unique name
+CLUSTER_NAME=e2etlslab
+
+# Cluster Creation Command
+az aks create \
+-g $RG \
+-n $CLUSTER_NAME \
+--node-count 2 \
+--enable-app-routing \
+--app-routing-default-nginx-controller Internal \
+--enable-addons azure-keyvault-secrets-provider \
+--vnet-subnet-id $VNET_SUBNET_ID
+
+# Get the cluster identity
+CLUSTER_IDENTITY=$(az aks show -g $RG -n $CLUSTER_NAME -o tsv --query identity.principalId)
+
+# Get credentials
+az aks get-credentials -g $RG -n $CLUSTER_NAME
+```
+
+## Create the Azure Key Vault and Upload Certificate
+
+Now, let's create our Azure Key Vault and then create and upload our certificate.
+
+```bash
+# Create a key vault name
+KEY_VAULT_NAME=e2elab$RANDOM
+
+# Create the key value
+az keyvault create --name $KEY_VAULT_NAME --resource-group $RG --location $LOC --enable-rbac-authorization false
+```
+
+For Azure Front door I'll need a public certificate for my hostname. If I were doing this for a company, I would reach out to the team that owns public certificates, but in my case, I'll just use [Certbot](https://certbot.eff.org/) to generate a free certificate from [LetsEncrypt](https://letsencrypt.org/).
+
+I'm not going to get into all the specifics of using CertBot with LetsEncrypt, but the basics are as follows. The domain I'll be using is my 'crashoverride.nyc' domain.
+
+1. Get an internet reachable host capable of running a webserver on ports 80 and 443
+2. Create an A-record for your target domain to the public IP of the server. This is required for hostname validation used by Certbot
+3. Run the certbot command as a privileged user on that web server host mentioned in #1 above
+
+Here's a sample of the command I used to create the certificate:
+
+```bash
+sudo certbot certonly --key-type rsa --standalone -d e2elab.crashoverride.nyc
+```
+
+Certbot creates several files, all with the PEM file extension. This is misleading, as fullchain.pem is the 'crt' file and the privkey.pem is the 'key' file. To store these in Azure Key Vault as certs we'll need to package these files up in a PFX format.
```

Comment:
Might want to drop a link somewhere here to help users install certbot if they don't already have it. https://certbot.eff.org/instructions